src/main/java/by/exadel/internship/config/jwt/AuthTokenFilter.java

```diff
@@ -0,0 +1,63 @@
+package by.exadel.internship.config.jwt;
+
+import by.exadel.internship.service.UserDetailsServiceImpl;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.MDC;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class AuthTokenFilter extends OncePerRequestFilter {
+
+    private final JwtService jwtService;
+    private final String AUTHORIZATION = "Authorization";
+    private final String BEARER = "Bearer ";
+
+    private final UserDetailsServiceImpl userDetailsService;
+
+    private String parseJwt(HttpServletRequest request) {
+        String headerAuth = request.getHeader(AUTHORIZATION);
+        if (!StringUtils.isEmpty(headerAuth) && headerAuth.startsWith(BEARER)) {
+            return StringUtils.substring(headerAuth, 7);
+        }
+        return null;
+    }
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest httpServletRequest,
+                                    HttpServletResponse httpServletResponse,
+                                    FilterChain filterChain) throws ServletException, IOException {
+        try {
+            String jwt = parseJwt(httpServletRequest);
+            if (jwt != null && jwtService.validateJwtToken(jwt)) {
+                String username = jwtService.getUserNameFromJwtToken(jwt);
+
+                UserDetails userDetails = userDetailsService.loadUserByUsername(username);
+                UsernamePasswordAuthenticationToken authenticationToken =
+                        new UsernamePasswordAuthenticationToken(userDetails, null, userDetails.getAuthorities());
+                authenticationToken.setDetails(new WebAuthenticationDetailsSource().buildDetails(httpServletRequest));
+
+                SecurityContextHolder.getContext().setAuthentication(authenticationToken);
+            }
+        }catch (Exception e) {
+            MDC.put("class: ", AuthTokenFilter.class.getSimpleName());
+            log.error("Authentication Error" + e.getMessage());
+        }
+        filterChain.doFilter(httpServletRequest, httpServletResponse);
```

Comment:
@Fenydude what test scenarios did you cover? 
Please, check that user get 401 error in case problem with authentication (invalid token, not found user, invalid password) and 403 problem with access control in case attempting to secured endpoint.

Reply:
We get this message when we try to get access with invalid token ({"path":"/internships","error":"Unauthorized","message":"Full authentication is required to access this resource","status":401})
And this if we try to get access without the required role 
({"timestamp":"2021-04-20T09:01:03.461+00:00","status":403,"error":"Forbidden","message":"","path":"/api/internships"})
Now i am using system AuthentificationException and its message, but i can change if we need

Reply:
Excellent!
I think there's enough to use AuthentificationException from the box. If UI needs additional info, we can create our own one.